metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -391,41 +393,60 @@ public void replay(UnregisterBrokerRecord record) {
-        }
+        replayRegistrationChange(record, record.id(), record.epoch(), Optional.of(true));
     }
 
     public void replay(UnfenceBrokerRecord record) {
-        int brokerId = record.id();
-        BrokerRegistration registration = brokerRegistrations.get(brokerId);
-        if (registration == null) {
+        replayRegistrationChange(record, record.id(), record.epoch(), Optional.of(false));
+    }
+
+    public void replay(BrokerRegistrationChangeRecord record) {
+        replayRegistrationChange(record, record.brokerId(), record.brokerEpoch(), getFencingChange(record.fenced()));
+    }
+
+    private Optional<Boolean> getFencingChange(byte value) {
+        switch (value) {
+            case -1:
+                return Optional.of(false);
+            case 0:
+                return Optional.empty();
+            case 1:
+                return Optional.of(true);
+            default:
+                throw new RuntimeException(String.format("Invalid value for fenced: %d", value));
+        }
+    }
+
+    private void replayRegistrationChange(
+        ApiMessage record,
+        int brokerId,
+        long brokerEpoch,
+        Optional<Boolean> fenced
+    ) {
+        BrokerRegistration curRegistration = brokerRegistrations.get(brokerId);
+        if (curRegistration == null) {
             throw new RuntimeException(String.format("Unable to replay %s: no broker " +
                 "registration found for that id", record.toString()));
-        } else if (registration.epoch() !=  record.epoch()) {
+        } else if (curRegistration.epoch() != brokerEpoch) {
             throw new RuntimeException(String.format("Unable to replay %s: no broker " +
                 "registration with that epoch found", record.toString()));
         } else {
-            if (heartbeatManager != null) heartbeatManager.register(brokerId, false);
-            brokerRegistrations.put(brokerId, registration.cloneWithFencing(false));
-            updateMetrics(registration, brokerRegistrations.get(brokerId));
-            log.info("Unfenced broker: {}", record);
-        }
-        if (readyBrokersFuture.isPresent()) {
-            if (readyBrokersFuture.get().check()) {
-                readyBrokersFuture.get().future.complete(null);
-                readyBrokersFuture = Optional.empty();
+            BrokerRegistration nextRegistration = curRegistration;
+            if (fenced.isPresent()) {
+                nextRegistration = nextRegistration.cloneWithFencing(fenced.get());
+            }
+            if (!curRegistration.equals(nextRegistration)) {
+                brokerRegistrations.put(brokerId, nextRegistration);
+                updateMetrics(curRegistration, nextRegistration);
+            } else {
+                log.info("Ignoring no-op registration change for {}", curRegistration);
+            }
+            if (heartbeatManager != null) heartbeatManager.register(brokerId, nextRegistration.fenced());
```

Comment:
When will the heartbeatManager be `null`? Is this only in tests?

Reply:
it seems `heartbeatManager` will be null in a standby controller.

Reply:
yes